evm/src/cpu/kernel/asm/core/precompiles/expmod.asm

```diff
@@ -1,3 +1,260 @@
+    DUP5
+    %add_const(96)
+    // stack: 96 + l_B, l_E, len, l_M, l_E, l_B, kexit_info
+    PUSH @SEGMENT_CALLDATA
+    GET_CONTEXT
+    // stack: ctx, @SEGMENT_CALLDATA, 96 + l_B, l_E, len, l_M, l_E, l_B, kexit_info
+    DUP5
+    // stack: e_loc=len, ctx, @SEGMENT_CALLDATA, 96 + l_B, l_E, len, l_M, l_E, l_B, kexit_info
+    PUSH @SEGMENT_KERNEL_GENERAL
+    PUSH 0
+    // stack: dst=(0, @SEGMENT_KERNEL_GENERAL, e_loc), src=(ctx, @SEGMENT_CALLDATA, 96 + l_B), l_E, len, l_M, l_E, l_B, kexit_info
+    %memcpy
+    // stack: len, l_M, l_E, l_B, kexit_info
+
+    // Copy M to kernel general memory.
+    DUP2
+    // stack: l_M, len, l_M, l_E, l_B, kexit_info
+    DUP5
+    DUP5
+    ADD
+    %add_const(96)
+    // stack: 96 + l_B + l_E, l_M, len, l_M, l_E, l_B, kexit_info
+    PUSH @SEGMENT_CALLDATA
+    GET_CONTEXT
+    // stack: ctx, @SEGMENT_CALLDATA, 96 + l_B + l_E, l_M, len, l_M, l_E, l_B, kexit_info
+    DUP5
+    %mul_const(2)
+    // stack: m_loc=2*len, ctx, @SEGMENT_CALLDATA, 96 + l_B + l_E, l_M, len, l_M, l_E, l_B, kexit_info
+    PUSH @SEGMENT_KERNEL_GENERAL
+    PUSH 0
+    // stack: dst=(0, @SEGMENT_KERNEL_GENERAL, m_loc), src=(ctx, @SEGMENT_CALLDATA, 96 + l_B + l_E), l_M, len, l_M, l_E, l_B, kexit_info
+    %memcpy
+    // stack: len, l_M, l_E, l_B, kexit_info
+
+    SWAP3
+    %pop3
+    // stack: len, kexit_info
+
+    PUSH expmod_contd
+    // stack: expmod_contd, len, kexit_info
+    DUP2
+    // stack: len, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(11)
+    // stack: s5=11*len, len, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(9)
+    // stack: s4=9*len, len, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(7)
+    // stack: s3=7*len, len, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(5)
+    // stack: s2=5*len, len, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(4)
+    // stack: s1=4*len, len, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(3)
+    // stack: out=3*len, len, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(2)
+    // stack: m_loc=2*len, len, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    PUSH 0
+    // stack: b_loc=0, e_loc=len, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    DUP2
+    // stack: len, b_loc, e_loc, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    %jump(modexp_bignum)
```

Comment:
The number on the stack are counting bytes, but they should be counting 128-bit limbs. For example, if the values are 256-bit, currently `len=32` (for 32 bytes) but it should be `len=2` (for 2 128-bit limbs).

Reply:
Ohh okay I fixed that, but a bigger issue is that the calldata is in bytes, whereas my `modexp` code expects its input in limbs — right? So I'll need to pack it before storing in kernel general memory?

Reply:
I added code to pack the inputs and unpack the outputs! It's kinda gnarly, a lot of lines, but I'm not sure how to do it differently. But let me know if you have suggestions to make it cleaner or more readable!